src/framework/ui/view/menuitem.cpp

```diff
@@ -0,0 +1,225 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include "menuitem.h"
+
+#include <QVariantMap>
+
+#include "shortcuts/shortcutstypes.h"
+
+using namespace mu::ui;
+
+MenuItem::MenuItem(const UiAction& a)
+{
+    m_id = QString::fromStdString(a.code);
+    m_action = a;
+}
+
+QString MenuItem::id() const
+{
+    return m_id;
+}
+
+QString MenuItem::section() const
+{
+    return m_section;
+}
+
+UiActionState MenuItem::state() const
+{
+    return m_state;
+}
+
+bool MenuItem::selectable() const
+{
+    return m_selectable;
+}
+
+bool MenuItem::selected() const
+{
+    return m_selected;
+}
+
+MenuItemRole MenuItem::role() const
+{
+    return m_role;
+}
+
+QList<MenuItem*> MenuItem::subitems() const
+{
+    return m_subitems;
+}
+
+UiAction MenuItem::action() const
+{
+    return m_action;
+}
+
+mu::actions::ActionData MenuItem::args() const
+{
+    return m_args;
+}
+
+bool MenuItem::isValid() const
+{
+    return !m_id.isEmpty();
+}
+
+QString MenuItem::shortcutsTitle() const
+{
+    return mu::shortcuts::sequencesToNativeText(m_action.shortcuts);
+}
+
+QString MenuItem::portableShortcut() const
+{
+    return QString::fromStdString(mu::shortcuts::Shortcut::sequencesToString(m_action.shortcuts));
+}
+
+void MenuItem::setId(const QString& id)
+{
+    if (m_id == id) {
+        return;
+    }
+
+    m_id = id;
+    emit idChanged(m_id);
+}
+
+void MenuItem::setSection(const QString& section)
+{
+    if (m_section == section) {
+        return;
+    }
+
+    m_section = section;
+    emit sectionChanged(m_section);
+}
+
+void MenuItem::setState(const UiActionState& state)
+{
+    if (m_state == state) {
+        return;
+    }
+
+    m_state = state;
+    emit stateChanged();
+}
+
+void MenuItem::setSelectable(bool selectable)
+{
+    if (m_selectable == selectable) {
+        return;
+    }
+
+    m_selectable = selectable;
+    emit selectableChanged(m_selectable);
+}
+
+void MenuItem::setSelected(bool selected)
+{
+    if (m_selected == selected) {
+        return;
+    }
+
+    m_selected = selected;
+    emit selectedChanged(m_selected);
+}
+
+void MenuItem::setRole(MenuItemRole role)
+{
+    if (m_role == role) {
+        return;
+    }
+
+    m_role = role;
+    emit roleChanged(role_property());
+}
+
+void MenuItem::setSubitems(const QList<MenuItem*>& subitems)
+{
+    if (m_subitems == subitems) {
+        return;
+    }
+
+    m_subitems = subitems;
+    emit subitemsChanged(m_subitems);
+}
+
+void MenuItem::setAction(const UiAction& action)
+{
```

Comment:
It would be nice to have a check for equivalent values here:

if (m_action == action) {
return;
}

m_action = action;
emit actionChanged();